### PR TITLE
fix for #4518: (no-trailing-whitespace) flags leading empty line as error

### PR DESCRIPTION
**Pull request by rrogowski**
#### PR checklist

- [x] Addresses an existing issue: #4518 
- [x] New feature, bugfix, or enhancement
  - [x] Includes tests
- [ ] Documentation update

#### Overview of change:

Temporary fix to handle files starting with a byte order mark (see https://github.com/ajafff/tsutils/issues/94).

#### CHANGELOG.md entry:

[bugfix] `no-trailing-whitespace` no longer flags files starting with a byte order mark

### CODE REVIEW

File: src/rules/noTrailingWhitespaceRule.ts [hunk 1]
```diff
@@ -87,7 +89,14 @@ function walk(ctx: Lint.WalkContext<Options>) {
     for (const line of getLineRanges(sourceFile)) {
         // \s matches any whitespace character (equal to [\r\n\t\f\v ])
         const match = text.substr(line.pos, line.contentLength).match(/\s+$/);
-        if (match !== null && !(ctx.options.ignoreBlankLines && match.index === 0)) {
+        if (
+            match !== null &&
+            !(ctx.options.ignoreBlankLines && match.index === 0) &&
+            // See https://github.com/palantir/tslint/issues/4518
+            // and https://github.com/ajafff/tsutils/issues/94
+            // TODO: Remove this check once tsutils handles Byte Order Markers correctly.
```
Comment (JoshuaKGoldberg) [inline src/rules/noTrailingWhitespaceRule.ts]: You can remove this TODO; see discussion in the tsutils issue.

File: src/rules/noTrailingWhitespaceRule.ts [hunk 2]
```diff
@@ -87,7 +89,14 @@ function walk(ctx: Lint.WalkContext<Options>) {
...
+            match[0].charCodeAt(0) !== ZERO_WIDTH_NO_BREAK_SPACE
```
Comment (JoshuaKGoldberg) [inline src/rules/noTrailingWhitespaceRule.ts]: Missing test case: what if the line starts with a BOM but also has trailing whitespace?

File: src/rules/noTrailingWhitespaceRule.ts (hunk 1 above)
Comment (rrogowski) [reply src/rules/noTrailingWhitespaceRule.ts]: Got it!

File: src/rules/noTrailingWhitespaceRule.ts (hunk 2 above)
Comment (rrogowski) [reply src/rules/noTrailingWhitespaceRule.ts]: Thanks, I've handled this case now!

**Comment by rrogowski** [comment]
@JoshuaKGoldberg suggestions have been addressed. Ready for your review.
